lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -391,13 +412,102 @@ class BasicBlockCloner : public SILClonerWithScopes<BasicBlockCloner> {
     return Value;
   }
 
-  // Update ValueMap so that occurrences of `oldValue` are replaced with
-  // `newValue` when cloning.
+  /// Update ValueMap so that occurrences of `oldValue` are replaced with
+  /// `newValue` when cloning.
   void updateValueMap(SILValue oldValue, SILValue newValue)  {
     auto emplaceResult = ValueMap.try_emplace(oldValue, newValue);
     assert(emplaceResult.second && "Updating the same key in ValueMap multiple "
                                    "times during SESE cloning.");
   }
+
+  /// Clone the body of `loop` starting from `startBlock` and nest the cloned
+  /// fragment into the parent loop. If `startBlock` is the same as the header
+  /// of `loop`, we clone the entire loop including the back edge.  Otherwise,
+  /// we clone one iteration of the loop body without the back edge.
+  SILLoop *cloneLoop(SILLoopInfo *LI, SILLoop *loop, SILBasicBlock *startBlock) {
```

Comment:
from the call site below, it seems `startBlock` is always `header`. Why do we need the other code path?

Reply:
The other code path will be used in a subsequent PR that I am preparing.

Reply:
@mhong has a valid point

Reply:
In this use case, we are only cloning part of the loop. The `header` that is being passed at the call-site is the original header of the loop before canonicalization. After canonicalization, the loop will look like this:
```
  new_header: 
     cond_br stayInLoop header loop_exit
  header:
     ...
     br latch
  latch: 
     ...
     br new_header 
  exit:
     ...
``` 
We are only cloning the part of the loop reachable from `header` and the result would look something like this (primed names are the clones): 
```
   header':
     ...
     br latch'
  latch': 
     ...
     br new_header 

  new_header: 
     cond_br stayInLoop header loop_exit
  header:
     ...
     br latch
  latch: 
     ...
     br new_header 
  exit:
     ...
```
In the other path, we will pass in `new_header` and thereby clone the entire loop. Will send out that PR shortly.